### ISSUE TITLE
beforeUpdate in User.js causes double hashing when calling .save()

### ISSUE THREAD

**Issue by willjleong** [ghaiklor/generator-sails-rest-api#70]
Updating the password should be a custom method called on the User model.  If .save() is called on a user instance, the password gets double hashed and user will be unable to login.  Another class or instance method should used only for updating the password.

**Comment by ghaiklor** [comment]
@willjleong how can I reproduce it? Just make `PUT /v1/user` and update the model?

**Comment by mhipo1364** [comment]
@willjleong  @ghaiklor  I think we can drop getting hashed password on beforeUpdate method, then use it in another custom method, because beforeUpdate and ... works with signals, and runs on each update

**Comment by ghaiklor** [comment]
@willjleong @mhipo1364 beforeCreate and beforeUpdate I'm checking if you sending new password in values. If not, then values.password is not updating. So I need to know how can I reproduce it.
We can do another method in User instance that will just receive new plain password and updating in model for new one, without any checks.

**Comment by willjleong** [comment]
If you do something custom in a controller where you add something to the user and the call .save() on it, the password will get hashed again, and the user will no  longer be able to login.  

For example if your user has many favorites relationship you'd do something like this:
```
User.findOne(req.user.id).then(function(user) {
    Favorite.create(req.allParams()).then(function(){
       user.add(favorite);
       user.save(res.ok);
    })
})
```

the .save() instance method will actually trigger the beforeUpdate(){} class hook and because the password is not being removed before update, the password gets hashed again.

**Comment by ghaiklor** [comment]
@willjleong I see, any ideas to fix it? Except creating another method in model. If I'll not find any solutions I make smth like

```javascript
// models/User.js
module.exports = {
  // ...
  refreshPassword: function(password) {
    this.password = HashService.bcrypt.hashSync(password);
    return this.save();
  }
};
```

**Comment by mhipo1364** [comment]
@ghaiklor I think keep it simple, changing password scenario in most popular web apps has a separated scenario, so you can change it in another action; just like your example.

**Comment by willjleong** [comment]
@ghaiklor I think that is a pretty good solution.  
Just have to make sure that when the user is updated with a param 'password' just call user.refreshPassword() in the User.update().then(function(){return user.refreshPassword(param.password) }).  

You'll probably have to override the update function in UserController.js. Also probably add a reset password route in AuthController.js that requires the isAuthenticated policy.

**Comment by ghaiklor** [comment]
@mhipo1364 @willjleong yeah, seems good solution. I'll start on it from tomorrow.

**Comment by synergycns** [comment]
@ghaiklor you could also either have beforeUpdate check to see if the password is less than 60 characters, or implement some sort of isHash(password):Boolean in your HashService

beforeUpdate: function (values, next) {
    if (values.password && values.password.length < 60) {
      values.password = HashService.bcrypt.hashSync(values.password);
      console.log('beforeUpdate - Hashed password: ' + values.password);
    }
    next();
  }

**Comment by mhipo1364** [comment]
@synergycns : What if user provides a password with more than 60 characters? :smiley:

**Comment by ghaiklor** [comment]
@mhipo1364 @synergycns yeah, I want to be more secured, so I have password with Russian, Ukrainian and English alphabets :smiley: 
But seriously, what's the reason for this case ? Have you any use-cases for it ?

**Comment by synergycns** [comment]
Bcrypt has a maximum input length of 56 bytes and generates hashes that are 60 characters long. So, if you're dealing with passwords that are longer than 55 characters, you can't use bcrypt anyway. Otherwise, the hash string will always be 60 characters in length and that's the length that I always use for database columns that will contain bcrypt-hashed strings. Also discussed in more details here: http://stackoverflow.com/questions/5881169/what-column-type-length-should-i-use-for-storing-a-bcrypt-hashed-password-in-a-d

**Comment by synergycns** [comment]
@ghaiklor just to make things a bit cleaner than just looking for a string that is less than 60 characters, I've made the following changes on my fork. Hope this helps:

1. HashService.js

var hashes = require('sails-service-hash');
var BCRYPT_REGEX = /^\$2[aby]\$[0-9]{2}\$.{53}$/;

module.exports = {
  bcrypt: hashes.create('bcrypt', {}),
  isBCryptHash: function(string) {
    var blnIsHash = BCRYPT_REGEX.test(string);
    return blnIsHash;
  }
};

2. User.js

beforeUpdate: function (values, next) {
    if(values.password) {
      var blnIsHash = HashService.isBCryptHash(values.password);
      if(!blnIsHash) {
        values.password = HashService.bcrypt.hashSync(values.password);
      }
    }
    next();
  }

**Comment by IncoCode** [comment]
```js
beforeUpdate: function (values, next) {
	if (values.id) {
		if (values.password) {
			return User.findOne({id: values.id})
			.then(function (user) {
				if (values.password === user.password) {
					return next();
				}
				values.password = encode pass;
				next();
			})
			.catch(next);
		}
	} else if (values.password) {
		values.password = encode pass;
	}

	next();
}
```

**Comment by synergycns** [comment]
Does the "findOne" call result in an additional DB query? I'm new to sails so, I don't know. If it does, than it seems extraneous to perform another DB lookup on the model prior to updating. If it doesn't than, yes, this way is better.

**Comment by komocode** [comment]
Need to watch out for users that provide passwords longer than 60 characters. Some use password managers and crank up the length to 64 randomized characters.

**Comment by ghaiklor** [comment]
@komocode what can we do with that though ? My password manager generates password up to 16 symbols and it's fine. All what can be done is respond with error that you have typed too long password.

**Comment by komocode** [comment]
Sorry, I didn't see that @mhipo1364 already mentioned the 60 characters. but yes we can put a check on password being >32 characters. If it is, it's a hash, if not, hash it.

**Comment by ghaiklor** [comment]
@willjleong what you think about this solution?

```javascript
  beforeUpdate: function (values, next) {
    var id = values.id;
    var password = values.password;

    if (id && password) {
      return User
        .findOne({id: id})
        .then(function (user) {
          if (password === user.password) {
            return next();
          } else {
            values.password = HashService.bcrypt.hashSync(password);
            return next();
          }
        })
        .catch(next);
    } else {
      next();
    }
  },

  beforeCreate: function (values, next) {
    values.password = HashService.bcrypt.hashSync(values.password);
    next();
  }
```

**Comment by ghaiklor** [comment]
We can check if password is bcrypt hash just with regex. It'll simplify the solution above.

**Comment by synergycns** [comment]
I already provided example code above to do exactly that so the extraneous DB lookup isn't needed :) I've used it in a few different components since posting it and it works great for me, at least for my applications, since my input passwords are always less than 72 characters. What I posted works for Bcrypt 2a, 2b, and 2y hashes (not regular old Bcrypt 2 since that produces a 59 character hash string although it can easily be made to by removing the "[aby]" portion of the regex string) with 2y being the most applicable since this is what HashService uses by default. The hashed strings produced by HashService are ALWAYS 60 characters long with a predefined structure and a maximum input password length of 72 characters. Feel free to use it if it helps.
